Islandora-Devops/claw-playbook
# Initial Docker impl

# Draft PR to demonstrate Islandora 8 in Docker

Creates and runs two Docker containers:

* An `ansible` container that, upon startup, runs ansible to install Islandora 8 to the `vm` container
* A `vm` container that plays the role of an empty Centos 7 box upon which Islandora 8 is installed via the `ansible` container.

## Motivation
* Providing the start of a development environment that would allow modeling of "services deployed on separate hosts" as containers.  
  * Right now, though, everything is installed to a single container whose role is analogous to the Vagrant VM.  It's a start, though.
* Because using Ansible + Vagrant + Virtualbox in Windows is possible, but impractically awkward

## To Test
* Make sure Docker is installed, as well as `docker-compose` (usually comes with the Docker distribution, e.g on Mac and Windows)
* Make sure the Islandora ports are free on your machine (e.g. there isn't another Islandora running on the same ports using Vagrant)
* Check out this PR, and in the repository's directory (`claw-playbook/`) do `docker-compose build`
* Do `docker-compose up -d`
* Watch the logs of the `ansible` container via ` docker logs -f ansible`.  That's the ansible output!  
* At this point, the only way to know it's done/ready is to look at the logs.  So watch the logs until Ansible is finished
  * Once the Ansible stops, the container will just run infinitely tailing /dev/null.  The container won't exit if it succeeds or fails.  It runs infinitely because that has been helpful for me in debugging Ansible issues.
* If Ansible succeeds, try logging into Islandora at the usual address `localhost:8000`, and poke around!  

## Tips and tricks

* You can hop onto any container while it's running by _docker exec -it <name> /bin/bash_.  For example, `docker exec -it vm /bin/bash`
* `docker-compose stop` will stop all containers, but keep state.  You can later on do a `docker-compose up vm` just to start the Islandora "vm" container. 
  * I haven't yet put the persisted state into volumes, so files/state are just written in the container.  As a result, all state disappears in a _poof_ once the containers are destroyed, e.g. via `docker-compose down`.  `docker-compose stop` preserves state.  

See if this works for you @dannylamb, or anybody else interested.  Have fun!
